### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by MusinaPolina**

### CODE REVIEW

File: src/jvmTest/kotlin/test.kt [hunk 1]
```diff
@@ -39,61 +57,127 @@ internal class DiamondTests {
...
+    fun test(filepath: String) {
+        assertEquals(true, runTest("$testsPath$filepath"))
     }
 
+    private val assumptionTestsPath = "src/jvmTest/resources/testCover/small"
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("getAssumptionFilenames")
+    fun assumptionTest(filepath: String) {
+        assertEquals(true, runAssumptionTest("$assumptionTestsPath$filepath"))
```
Comment (Lipen) [inline src/jvmTest/kotlin/test.kt:181]: It is unclean how `$assumptionTestsPath$filepath` works without a `/`-separator... It is very bold to assume `filepath` starts with `/`. So, I suggest either adding `/` between paths (note: `dirname//filename` is a perfectly valid path, despite or duplicate `/`) or using Kotlin's path-joining mechanisms (`path.resolve`, `Paths.get`, etc).

File: src/jvmTest/resources/DESCRIPTION.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+## Test description
```
Comment (Lipen) [inline src/jvmTest/resources/DESCRIPTION.md:1]: Description of tests is still incomplete and too scarce. Do not forget to improve it each time you add a benchmark!

File: src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt [hunk 3]
```diff
@@ -0,0 +1,282 @@
+package org.kosat.heuristics
+
+import org.kosat.CDCL
+import org.kosat.Clause
+import org.kosat.VarValue
+import org.kosat.variable
+
+// TODO: add docs and refactoring
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt:8]: TODO: add docs!
TODO: refactor this class!

File: src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+class Preprocessor(private val solver: CDCL) {
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt:9]: The cognitive complexity of this class is not negligible anymore... Do something with it! Simplify the code, add comments, etc.

File: src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    private val oldNumeration = MutableList(solver.numberOfVariables + 1) { index -> index }
+    private var startClauses = listOf<Clause>()
+    private var startOccurrence = listOf<MutableList<Int>>()
+    private val deletingOrder = mutableListOf<Int>()
+    private val isClauseDeleted = MutableList(solver.constraints.size) { false }
+    private val clauseLimit = 600
+    private val hash = LongArray(2 * solver.numberOfVariables + 1) { 1L.shl(it % 64) }
+
+    fun addClause(clause: Clause) {
+        clause.forEach { lit -> litOccurrence[variable(lit)].add(solver.constraints.lastIndex) }
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/heuristics/Preprocessor.kt:20]: Why not just use good-old for-loop?
```kotlin
for (lit in clause.lits) {
    litOccurrence[variable(lit)].add(solver.constraints.lastIndex)
}
```

File: src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt [hunk 6]
```diff
@@ -94,30 +96,33 @@ class PriorityQueue {
     }
 
     fun buildHeap(activity: MutableList<Double>) {
-        for (ind in 1..activity.lastIndex) {
+        for (ind in 0..activity.lastIndex) {
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt:99]: Consider
```kotlin
for (ind in activity.indices) {...}
```
or even
```kotlin
for (ind_act in activity.withIndex()) {
    heap.add(ind_act)
}
```
to eliminate index-related bugs.

File: src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt [hunk 7]
```diff
@@ -94,30 +96,33 @@ class PriorityQueue {
     }
 
     fun buildHeap(activity: MutableList<Double>) {
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt:98]: I don't really like this implementation. The standard way of build a heap from an array is as simple as follows:
1. Allocate a dummy heap from the array. (`heap = array`, index/order/whatever structure use use for indexing/ordering = initially empty vec).
2. Iterate over elements of `heap` from the end(!). Sift_down (w.r.t. min-heap) each element.
3. ??? Done!

File: src/commonMain/kotlin/org/kosat/heuristics/VariableSelector.kt [hunk 8]
```diff
@@ -101,7 +109,7 @@ class VSIDS(private var numberOfVariables: Int = 0) : VariableSelector() {
                 break
             }
         }
-        return v
+        return v * 2
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/heuristics/VariableSelector.kt:112]: Why?
Are you trying to convert `v` to a "literal" (positive by default)? In this case, your function name is quite wrong: you promised "`get...Variable`", but returning literal.
You already know the solution: wrap all variables and literals with `Var` and `Lit` inline-classes in order to eliminate the obscurity of `Int`s.

If you are doing `2*v` for any other reason, then please explain!

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 9]
```diff
@@ -1,9 +1,6 @@
 package org.kosat
 
-import org.kosat.heuristics.Preprocessor
-import org.kosat.heuristics.Restarter
-import org.kosat.heuristics.VariableSelector
-import org.kosat.heuristics.VsidsWithoutQueue
+import org.kosat.heuristics.*
```
Comment (Lipen) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:3]: Star-imports again!

File: src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            heap[ind] = Pair(heap[ind].first / divisor, heap[ind].second)
+        }
+    }
+
+    fun getMax(): Pair<Double, Int> {
+        require(size != 0)
+        return heap[0]
+    }
+
+    fun deleteMax() {
```
[the remark this answers is not in the document]
Comment (d-aver) [reply src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt]: ok

File: src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        swap(0, size - 1)
+        order[heap[size - 1].second] = -1
+        size--
+        if (heap.isNotEmpty()) {
+            heapify(0)
+        }
+    }
+
+    // if some value of vertex increased this function lift this vertex up to save heap structure
+    fun liftVertex(u: Int) {
```
[the remark this answers is not in the document]
Comment (d-aver) [reply src/commonMain/kotlin/org/kosat/heuristics/PriorityQueue.kt]: ok

**Pull request merged**